### PR TITLE
Support for HTTPS TLS connections on older devices

### PR DESCRIPTION
**Pull request by marcopar**
Devices with SDK version between 16 and 20 (and even some on 21) support TLS but TLS is not enabled by default.
Many websites are switching to HTTPS requiring the use of TLS (i.e. NASA sites)

This patch adds a factory for creating OkHttpClient with TLS enabled on those SDK versions.
It then modifies DownloadArtTask so it uses that factory.

Without this patch Muzei fails to download images from those websites on those devices because with TLS disabled the SSL handshake fails.

Background:
For example it fails to download images using my muzei source that fetches images from the NASA / SDO project website that switched to TLS at the end of 2016.
https://github.com/marcopar/SDOViewer
It is possible to reproduce the problem running Muzei and my source on a API19 emulator instance.

This patch has been built following the indications on this OkHttp tracker:
https://github.com/square/okhttp/issues/2372
As you can see the issue is referred by several other trackers recognizing the problem.

### CODE REVIEW

**Comment by marcopar** [comment]
I adapted the implementation to somewhat match the one included in Facebook React, that is based on the same source but i think it's more polished and enables support also for TLS 1.1.
See:
https://github.com/facebook/react-native/pull/9840

Bonus: If they find some bug it will be easier to merge the fixes.

File: main/src/main/java/com/google/android/apps/muzei/TLSSocketFactory.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
+package com.google.android.apps.muzei;
+
+/**
+ * Created by marcopar on 11/02/17.
```
Comment (ianhanniballake) [inline main/src/main/java/com/google/android/apps/muzei/TLSSocketFactory.java]: Add the copyright/licensing block please

Comment (ianhanniballake) [inline main/src/main/java/com/google/android/apps/muzei/TLSSocketFactory.java]: Remove this comment block

File: main/src/main/java/com/google/android/apps/muzei/OkHttpClientFactory.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
+package com.google.android.apps.muzei;
```
Comment (ianhanniballake) [inline main/src/main/java/com/google/android/apps/muzei/OkHttpClientFactory.java]: Add the copyright/licensing block please

File: main/src/main/java/com/google/android/apps/muzei/OkHttpClientFactory.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+import javax.net.ssl.X509TrustManager;
+
+import okhttp3.ConnectionSpec;
+import okhttp3.OkHttpClient;
+import okhttp3.TlsVersion;
+
+/**
+ * Factory for OkHttpClient, supports the creation of clients enabling TLS on devices where it's not enabled by default (mainly pre lollipop)
+ */
+
```
Comment (ianhanniballake) [inline main/src/main/java/com/google/android/apps/muzei/OkHttpClientFactory.java]: Remove this extra line

File: main/src/main/java/com/google/android/apps/muzei/OkHttpClientFactory.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            specs.add(ConnectionSpec.CLEARTEXT);
+
+            client.connectionSpecs(specs);
+        } catch (Exception exc) {
+            Log.e(TAG, "Error while setting TLS", exc);
+        }
+
+        return client;
+    }
+
```
Comment (ianhanniballake) [inline main/src/main/java/com/google/android/apps/muzei/OkHttpClientFactory.java]: Remove extra lines

File: main/src/main/java/com/google/android/apps/muzei/OkHttpClientFactory.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+     */
+    public static OkHttpClient getNewOkHttpsSafeClient() {
+        return getNewOkHttpClient(isTLSEnableNeeded());
+    }
+
+    /**
+     * True if enabling TLS is needed on current device (SDK version >= 16 and < 22)
+     */
+    public static boolean isTLSEnableNeeded() {
+        if (Build.VERSION.SDK_INT >= 16 && Build.VERSION.SDK_INT < 22) {
```
Comment (ianhanniballake) [inline main/src/main/java/com/google/android/apps/muzei/OkHttpClientFactory.java]: Please use the Build.VERSION_CODES constants

**Comment by marcopar** [comment]
Request changes pushed.

**Comment by ianhanniballake** [comment]
Thanks for your contribution!

**Pull request merged**